src/db.c

```diff
@@ -1431,14 +1437,30 @@ dictEntry *dbFind(redisDb *db, void *key, dbKeyType keyType){
  *    it performs a dictScan over the appropriate `keyType` dictionary of `db`.
  * 3. If the slot is entirely scanned i.e. the cursor has reached 0, the next non empty slot is discovered. 
  *    The slot information is embedded into the cursor and returned.
+ *
+ * To restrict the scan to a single cluster slot, pass a valid slot as
+ * 'onlyslot', otherwise pass -1.
  */
-unsigned long long dbScan(redisDb *db, dbKeyType keyType, unsigned long long v, dictScanFunction *fn, int (dictScanValidFunction)(dict *d), void *privdata) {
+unsigned long long dbScan(redisDb *db, dbKeyType keyType, unsigned long long v,
+                          int onlyslot, dictScanFunction *fn,
+                          int (dictScanValidFunction)(dict *d), void *privdata) {
     dict *d;
     unsigned long long cursor = 0;
     /* During main dictionary traversal in cluster mode, 48 lower bits in the cursor are used for positioning in the HT.
      * Following 14 bits are used for the slot number, ranging from 0 to 2^14-1.
      * Slot is always 0 at the start of iteration and can be incremented only in cluster mode. */
     int slot = getAndClearSlotIdFromCursor(&v);
+    if (onlyslot >= 0) {
+        if (slot < onlyslot) {
+            /* Fast-forward to onlyslot. */
+            serverAssert(onlyslot < CLUSTER_SLOTS);
+            slot = onlyslot;
+            v = 0;
+        } else if (slot > onlyslot) {
+            /* The cursor is already past onlyslot. */
+            return 0;
+        }
+    }
```

Comment:
Given this `onlyslot` logic is only applicable for user db scan feature, my initial thought was we would handle this logic in the `scanGenericCommand` flow and send in a cursor with the appropriate slot. With this, we wouldn't need to update the method signature and pass around `-1` from other users of this method. WDYT ?

Reply:
That's true, but on the other hand, I wanted to avoid the unnecessary call to dbGetNextNonEmptySlot after scanning the wanted slot and to avoid decode and re-encode the cursor before the call to dbScan and also decode the cursor after the call to dbScan for the stop condition of the while loop.

So there are pros and cons of both approaches. I think adding the -1 is less complex actually, but maybe you can prove me wrong?

Reply:
Right, this avoids an unnecessary call to `dbGetNextNonEmptySlot`. I couldn't think of anything better. Please mark it as resolved (somehow I can't do it).